sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala

```diff
@@ -46,15 +48,24 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
   /**
    * An execution strategy for rules that indicates the maximum number of executions. If the
    * execution reaches fix point (i.e. converge) before maxIterations, it will stop.
+   * If throwsExceptionUponMaxIterations is equal to true, it will issue an exception
+   * TreeNodeException.
    */
-  abstract class Strategy { def maxIterations: Int }
+  abstract class Strategy {
+    def maxIterations: Int
+    def throwsExceptionUponMaxIterations: Boolean
+  }
 
   /** A strategy that only runs once. */
-  case object Once extends Strategy { val maxIterations = 1 }
+  case object Once extends Strategy {
+    override val maxIterations = 1
+    override val throwsExceptionUponMaxIterations = false
```

Comment:
@srowen This is used in multiple components, e.g., `Analyzer` and `Optimizer`. I am not sure how to do it in a more efficient way. Could you give me some hints how we did in the Spark? Any example in the existing code base? Thanks!